game-core/src/test/java/games/strategy/triplea/delegate/MatchesTest.java

```diff
@@ -254,4 +266,179 @@ void shouldNotMatchWhenSeaTerritoryIsUnownedAndDoesNotHaveTerritoryAttachment()
       assertThat(newMatch(), notMatches(seaTerritory));
     }
   }
+
+  @Nested
+  @ExtendWith(MockitoExtension.class)
+  final class UnitCanBeInBattle {
+
+    @Mock GamePlayer player;
+    GameData gameData;
+
+    @BeforeEach
+    void setupGameData() {
+      gameData = givenGameData().build();
+    }
+
+    @Test
+    void infrastructureShouldNormallyNotBeInBattle() {
+      when(gameData.getDiceSides()).thenReturn(6);
```

Comment:
Does this method need the other mocked elements of `gameData`, or do we mostly only need for the dice sides?

I'd recommend to try and fully mock objects within the test method where reasonable, at least within a test so the test is more self contained. For example if another test required slightly different mocking of a game data, we'd suddenly have a conflict between that test and any other that depended on the mock.

More fundamentally it probably is better to break up the data needed from game data, use only that data, then you can pass those data objects as test parameters. For example, rather than passing a game data to get dice sides to another method, instead pass the dice sides value directly.

Reply:
To run, the test needs other aspects of gameData.  That's because `Matches#unitTypeCanBeInBattle` will call `Matches#unitTypeIsSupporterOrHasCombatAbility` which in turn calls `UnitAttachment#getAttack` which will then call `TechAbilityAttachment#getAttackBonus` with the gameData that is on the UnitAttachment.  And it is this gameData that it will then look for tech abilities as well as check the dice sides.

Reply:
Okay, I think we are then already started down a path where we have a 'god mock' of a 'god class'. Let's watch out for this going forward. I think the thing for us to do more of us avoid passing game data and instead pass the data that is needed instead. For example, if units did not have a reference to game data, it would be far easier to instantiate them and use in testing.

Reply:
I did a quick test where I removed the `extend` in `Unit` so that it doesn't have a `GameData`.  There were not as many compile errors as I would expect.

Do you have a problem with giving `Unit` a reference to the `GameProperties`?  That is what most of the errors are related to.

Reply:
That is on the right path. Where I'd like to get is that we have one object called a "MapData" or something like that, that is just the data parsed from a map. "MapData" would then be an initial state. A "GameData" would be the final state of the game. A "GameData" would be a "MapData" plus N changes, if a game is new, the number of changes is zero and "GameData" and "MapData" would be equivalent.

With that, we need to think a bit about how the core data model is going to be structured.

In the short term, removing GameData reference for GameProperties sounds like the right path.

Though, does that work for serialization? IIRC the game data reference is to mark/enable the object for serialization into save games.